### PR TITLE
fix: prevent unicode truncation panics with non-ASCII input

### LINKED ISSUE #55: [MEDIUM] Unicode string truncation causes panics with non-ASCII input
## Summary

Multiple locations in the codebase truncate strings using byte-index slicing (e.g., `&s[..50]`). This panics at runtime when the byte index falls within a multi-byte UTF-8 character, such as emoji, CJK characters, or Arabic text. In the gateway context, this is a denial-of-service vector exploitable by any user sending non-ASCII messages.

**CWE:** [CWE-119: Improper Restriction of Operations within the Bounds of a Memory Buffer](https://cwe.mitre.org/data/definitions/119.html) (results in panic / denial of service)

## Affected Code

**Location 1 — Gateway WhatsApp handler:**

`src/gateway/mod.rs:376-379`
```rust
if msg.content.len() > 50 {
    format!("{}...", &msg.content[..50])  // PANIC if byte 50 is mid-character
} else {
    msg.content.clone()
}
```

**Location 2 — Agent loop (single-shot mode):**

`src/agent/loop_.rs:153-154`
```rust
let summary = if response.len() > 100 {
    format!("{}...", &response[..100])  // PANIC if byte 100 is mid-character
```

**Location 3 — Agent loop (interactive mode):**

`src/agent/loop_.rs:196-197`
```rust
let summary = if response.len() > 100 {
    format!("{}...", &response[..100])  // PANIC if byte 100 is mid-character
```

### Gateway panic (Location 1):
1. Deploy ZeroClaw with WhatsApp channel configured.
2. Send a WhatsApp message containing 13+ CJK characters (each is 3 bytes in UTF-8, so 13 chars = 39 bytes, 17 chars = 51 bytes):
   ```
   This message will panic: 这是一个测试消息用来触发崩溃的中文
   ```
3. The gateway handler panics with: `byte index 50 is not a char boundary; it is inside '触' (bytes 48..51)`

### Agent loop panic (Locations 2 & 3):
1. Start ZeroClaw in interactive mode.
2. Send a query that produces an LLM response containing non-ASCII characters near the 100-byte boundary.
3. If `auto_save` is enabled, the truncation code panics.

### Minimal reproduction:
```rust
fn main() {
    let s = "a]".repeat(30); // 30 emoji, each 4 bytes = 120 bytes
    println!("{}", &s[..50]); // PANIC: byte index 50 is not a char boundary
}
```

## Impact

- **Denial of service:** Any WhatsApp user sending non-ASCII text (extremely common in non-English locales) can crash the gateway request handler.
- **Session termination:** In the agent loop, LLM responses containing non-ASCII text can terminate the interactive session.
- **Reliability:** This affects all non-English users and any conversation involving emoji, which is the majority of real-world usage.

## Suggested Fix

Extract a reusable safe truncation helper and use it in all three locations:

```rust
/// Truncate a string to at most `max_chars` characters, appending "..." if truncated.
fn truncate_with_ellipsis(s: &str, max_chars: usize) -> String {
    match s.char_indices().nth(max_chars) {
        Some((idx, _)) => format!("{}...", &s[..idx]),
        None => s.to_string(),
    }
}
```

Replace all three locations:
```rust
// Before (all 3 locations):
if msg.content.len() > 50 { format!("{}...", &msg.content[..50]) } else { msg.content.clone() }

// After:
truncate_with_ellipsis(&msg.content, 50)
```

Place the helper in a shared utility module (e.g., `src/util.rs`) since it's used across `gateway` and `agent` modules.

### PR Requirements
- Create `truncate_with_ellipsis` helper function
- Replace all 3 unsafe truncation sites
- Add tests with:
  - Pure ASCII strings (no truncation needed, truncation needed)
  - Multi-byte UTF-8 strings (emoji, CJK, accented characters)
  - Strings exactly at the boundary length
  - Empty strings
- All existing tests must pass (`cargo test --verbose`)

### PR DESCRIPTION
**Pull request by fettpl**
## Summary
Replace unsafe byte-index string slicing (`&s[..50]`, `&s[..100]`) with a char-boundary-aware `truncate_with_ellipsis` helper. The old code panics when the truncation byte index falls within a multi-byte UTF-8 character (emoji, CJK, Arabic, etc.).

## Changes
- **New:** `src/util.rs` — `truncate_with_ellipsis(s, max_chars)` helper with 8 unit tests
- **Fixed:** `src/gateway/mod.rs:376` — WhatsApp message logging truncation
- **Fixed:** `src/agent/loop_.rs:153, 196` — response summary truncation (2 sites)
- **Bonus:** fix pre-existing clippy warnings (`is_multiple_of`, `too_many_lines` on browser tool)

## Test plan
- [x] `cargo test --verbose` — all tests pass (including 8 new tests for the helper)
- [x] `cargo clippy -- -D warnings` — clean
- [ ] Manual: send emoji/CJK text via WhatsApp webhook, verify no panic

Closes #55

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Refactor**
  * Centralized string truncation into a shared utility and replaced scattered manual truncation logic.

* **Chores**
  * Added lint suppression for long methods to reduce false positives and keep code style consistent.

* **Tests**
  * Added comprehensive unit tests for the new truncation behavior (including multibyte and boundary cases).

### CODE REVIEW

File: src/util.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+/// Truncate a string to at most `max_chars` characters, appending "..." if truncated.
```
[the remark this answers is not in the document]
Comment (fettpl) [reply src/util.rs]: Fixed in 35716fb — updated the doc comment to clarify that `max_chars` is the prefix length, not the total output length.

**Review by fettpl — COMMENTED** [review]
Both issues addressed in 35716fb: fixed the second truncation site in agent/loop_.rs (interactive mode, lines 193-197) and corrected the doc comment for `truncate_with_ellipsis`.

**Comment by theonlyhennygod** [comment]
Validated in isolated worktree. Local cargo clippy -- -D warnings: PASS. Full test suite (cargo test -- --test-threads=1): PASS. PR scope: Unicode-safe truncation helper and truncation panic fixes. Files changed: src/agent/loop_.rs, src/gateway/mod.rs, src/main.rs, src/security/secrets.rs, src/tools/browser.rs, src/util.rs. Closing per maintainer request.
